**ML lab1**

Выполнила: Машрабова Аида

**1. Загрузка данных**

In [2]:
import pandas as pd

In [3]:
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')

In [15]:
train.head()

ApplicationDate   Age  AnnualIncome  CreditScore  LoanAmount  LoanDuration  \
0      2010-06-26  27.0       66829.0        549.0     17290.0          60.0   
1      1996-09-23  55.0      172147.0        850.0     16110.0          36.0   
2      2015-01-19  51.0      300000.0        850.0     38436.0          36.0   
3      1981-05-12  25.0       34683.0        847.0     19186.0          48.0   
4      1995-05-07  55.0      300000.0        850.0     30437.0          48.0   

  MaritalStatus  NumberOfDependents HomeOwnershipStatus  MonthlyDebtPayments  \
0      Divorced                 1.0                Rent               1095.0   
1       Widowed                 1.0            Mortgage                211.0   
2       Married                 0.0            Mortgage                546.0   
3       Married                 0.0               Other                153.0   
4        Single                 2.0                Rent                562.0   

   ...  EmploymentStatus  EducationLevel  Experience  NetWorth  \
0  ...          Employed       Associate         4.0   35067.0   
1  ...          Employed     High School        33.0   27001.0   
2  ...          Employed        Bachelor        28.0  278382.0   
3  ...          Employed     High School         0.0    9224.0   
4  ...          Employed        Bachelor        31.0    4502.0   

   BaseInterestRate InterestRate  MonthlyLoanPayment  TotalDebtToIncomeRatio  \
0          0.257790     0.251465          508.970230                0.288013   
1          0.086110     0.093173          514.675859                0.050585   
2          0.108436     0.115443         1268.276385                0.072571   
3          0.100686     0.112822          498.505187                0.225415   
4          0.110437     0.089037          756.035156                0.052721   

   LoanApproved  RiskScore  
0           0.0  66.176500  
1           1.0  28.495737  
2           1.0  34.488104  
3           1.0  36.910753  
4           1.0  31.347091  

[5 rows x 36 columns]

In [16]:
test.head()

ID ApplicationDate  Age  AnnualIncome  CreditScore  LoanAmount  \
0   0      1979-01-03   49        207095          850       14458   
1   1      2006-04-10   28         64653          674       65491   
2   2      2013-11-28   41        300000          768       16811   
3   3      1995-08-24   50        300000          850       48737   
4   4      1980-06-15   35        300000          672       51916   

   LoanDuration MaritalStatus  NumberOfDependents HomeOwnershipStatus  ...  \
0            60        Single                   0                Rent  ...   
1            48        Single                   2                 Own  ...   
2            60        Single                   1                 Own  ...   
3            36      Divorced                   0            Mortgage  ...   
4            60        Single                   6            Mortgage  ...   

   JobTenure  EmploymentStatus  EducationLevel  Experience  NetWorth  \
0          6          Employed     High School          23     10626   
1          3          Employed     High School           3    299024   
2          3          Employed     High School          19      6979   
3          3     Self-Employed        Bachelor          26      8438   
4          7          Employed       Associate          12    436216   

   BaseInterestRate InterestRate  MonthlyLoanPayment  TotalDebtToIncomeRatio  \
0          0.104458     0.125350          325.532336                0.040650   
1          0.233491     0.260561         2210.247757                0.454410   
2          0.147811     0.154571          403.977431                0.034919   
3          0.118737     0.099712         1571.946906                0.085558   
4          0.230916     0.237727         1486.675235                0.063707   

   LoanApproved  
0             1  
1             0  
2             1  
3             1  
4             1  

[5 rows x 36 columns]

**2. EDA**

**2.1.** Размеры данных

In [4]:
print(f"Размер train.csv: {train.shape}")
print(f"Размер test.csv: {test.shape}")

Размер train.csv: (11017, 36)
Размер test.csv: (5000, 36)


**2.2.** Общая информация о данных

In [9]:
print("\ntrain.csv:\n")
print(train.info())
print("\ntest.csv:\n")
print(test.info())



train.csv:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11017 entries, 0 to 11016
Data columns (total 36 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   ApplicationDate             10487 non-null  object 
 1   Age                         10487 non-null  float64
 2   AnnualIncome                10487 non-null  float64
 3   CreditScore                 9986 non-null   float64
 4   LoanAmount                  9986 non-null   float64
 5   LoanDuration                10487 non-null  float64
 6   MaritalStatus               10487 non-null  object 
 7   NumberOfDependents          10487 non-null  float64
 8   HomeOwnershipStatus         10487 non-null  object 
 9   MonthlyDebtPayments         9986 non-null   float64
 10  CreditCardUtilizationRate   10487 non-null  float64
 11  NumberOfOpenCreditLines     10487 non-null  float64
 12  NumberOfCreditInquiries     10487 non-null  float64
 13  DebtToIncomeRatio 

В train.csv большинство столбцов имеют тип float64, а также присутствуют столбцы типа object. В test.csv схожая структура, но типы данных отличаются (например, возраст представлен как int64, а не float64).

**2.3.** Анализ пропущенных данных

In [10]:
print("\nПропущенные значения в train.csv:\n")
print(train.isnull().sum())
print("\nПропущенные значения в test.csv:\n")
print(test.isnull().sum())


Пропущенные значения в train.csv:

ApplicationDate                530
Age                            530
AnnualIncome                   530
CreditScore                   1031
LoanAmount                    1031
LoanDuration                   530
MaritalStatus                  530
NumberOfDependents             530
HomeOwnershipStatus            530
MonthlyDebtPayments           1031
CreditCardUtilizationRate      530
NumberOfOpenCreditLines        530
NumberOfCreditInquiries        530
DebtToIncomeRatio              530
BankruptcyHistory             1031
LoanPurpose                   1031
PreviousLoanDefaults           530
PaymentHistory                 530
LengthOfCreditHistory          530
SavingsAccountBalance          530
CheckingAccountBalance        1031
TotalAssets                   1031
TotalLiabilities               530
MonthlyIncome                  530
UtilityBillsPaymentHistory     530
JobTenure                      530
EmploymentStatus               530
EducationLevel     

В train.csv обнаружены пропущенные значения в ряде столбцов, а в test.csv пропуски отсутствуют.


In [26]:
missing_columns = train.columns[train.isnull().any()].tolist()
missing_data = train[missing_columns]
missing_pattern = missing_data.isnull().sum(axis=1).value_counts().sort_index()

print("Распределение количества пропусков в строках:")
print(missing_pattern)

Распределение количества пропусков в строках:
0.0    10487
Name: count, dtype: int64


In [22]:
print("Пропуски в RiskScore:", train['RiskScore'].isnull().sum())

Пропуски в RiskScore: 530


Удаляем 530 строк с пропущенным значением RiskScore, так как целевая переменная обязательна и эти данные не могут быть использованы в процессе предсказания.

In [23]:
train = train.dropna(subset=['RiskScore'])
print(f"Размер данных после удаления пропусков в RiskScore: {train.shape}")
print(f"Оставшиеся пропуски в RiskScore: {train['RiskScore'].isnull().sum()}")

Размер данных после удаления пропусков в RiskScore: (10487, 37)
Оставшиеся пропуски в RiskScore: 0


После удаления строк с RiskScore осталось 501 строка с 9 пропусками в других столбцах Заполняем пропуски медианами и модами.

In [25]:
numeric_columns = train.select_dtypes(include=['float64', 'int64']).columns
train[numeric_columns] = train[numeric_columns].fillna(train[numeric_columns].median())
categorical_columns = train.select_dtypes(include='object').columns
for col in categorical_columns:
    mode_val = train[col].mode()[0]
    train[col] = train[col].fillna(mode_val)
print(f"Оставшиеся пропуски после заполнения: {train.isnull().sum().sum()}")

Оставшиеся пропуски после заполнения: 0
